src/Operation/BulkWrite.php

```diff
@@ -341,28 +341,30 @@ public function execute(Server $server)
                     break;
 
                 case self::INSERT_ONE:
-                    $insertedDocument = isset($this->options['codec'])
-                        ? $this->options['codec']->encodeIfSupported($args[0])
-                        : $args[0];
-                    // Psalm's assert-if-true annotation does not work with unions, so
-                    // assert the type manually instead of using is_document
-                    // See https://github.com/vimeo/psalm/issues/6831
-                    assert(is_array($insertedDocument) || is_object($insertedDocument));
-
-                    $insertedIds[$i] = $bulk->insert($insertedDocument);
+                    if (isset($this->options['codec'])) {
+                        $args[0] = $this->options['codec']->encodeIfSupported($args[0]);
+
+                        // Psalm's assert-if-true annotation does not work with unions, so
+                        // assert the type manually instead of using is_document
+                        // See https://github.com/vimeo/psalm/issues/6831
+                        assert(is_array($args[0]) || is_object($args[0]));
```

Comment:
Should we worry about PackedArray here? `is_document()` does:

```
is_array($document) || (is_object($document) && ! $document instanceof PackedArray)
```

My gut says, "Probably?", since the intention here was to inline `is_document()` for Psalm's benefit.

Before addressing this, consider my comment below for `REPLACE_ONE`, though.